src/expertsystem/reaction/__init__.py

```diff
@@ -262,87 +550,86 @@ def set_allowed_interaction_types(
-                [
-                    (seed_graph, x)
-                    for x in self._determine_graph_settings(seed_graph)
-                ]
-            )
+    def create_problem_sets(self) -> Dict[float, List[ProblemSet]]:
+        topology_graphs = self.__build_topologies()
+        problem_sets = []
+        for topology in topology_graphs:
+            for initial_facts in self.__create_initial_facts(topology):
+                problem_sets.extend(
+                    [
+                        ProblemSet(
+                            topology=topology,
+                            initial_facts=_convert_to_graph_props(
+                                initial_facts
+                            ),
+                            solving_settings=x,
+                        )
+                        for x in self.__determine_graph_settings(
+                            topology, initial_facts
+                        )
+                    ]
+                )
         # create groups of settings ordered by "probability"
-        graph_settings_groups = group_by_strength(graph_setting_pairs)
-        return graph_settings_groups
+        return _group_by_strength(problem_sets)
 
-    def _build_topologies(self) -> List[Topology]:
+    def __build_topologies(self) -> List[Topology]:
         all_graphs = self.topology_builder.build_graphs(
             len(self.initial_state), len(self.final_state)
         )
         logging.info(f"number of topology graphs: {len(all_graphs)}")
         return all_graphs
 
-    def _create_seed_graphs(
-        self, topology_graphs: List[Topology]
-    ) -> List[StateTransitionGraph[ParticleWithSpin]]:
+    def __create_initial_facts(self, topology: Topology) -> List[InitialFacts]:
         # initialize the graph edges (initial and final state)
-        init_graphs: List[StateTransitionGraph[ParticleWithSpin]] = []
-        for topology_graph in topology_graphs:
-            initialized_graphs = initialize_graph(
-                topology=topology_graph,
-                particles=self.__particles,
-                initial_state=self.initial_state,
-                final_state=self.final_state,
-                final_state_groupings=self.final_state_groupings,
-            )
-            init_graphs.extend(initialized_graphs)
-        for graph in init_graphs:
-            graph.graph_node_properties_comparator = (
-                CompareGraphNodePropertiesFunctor()
-            )
 
-        logging.info(f"initialized {len(init_graphs)} graphs!")
-        return init_graphs
+        initial_facts = create_initial_facts(
+            topology=topology,
+            particles=self.__particles,
+            initial_state=self.initial_state,
+            final_state=self.final_state,
+            final_state_groupings=self.final_state_groupings,
+        )
+
+        logging.info(f"initialized {len(initial_facts)} graphs!")
+        return [InitialFacts(edge_props=x) for x in initial_facts]
```

Comment:
Shouldn't `create_initial_facts` return an instance of `InitialFacts`? In other words, why is `InitialFacts` defined under `reaction.__init__` and not under `combinatorics`?

Reply:
Good point. e263774e6ad99f0965b734037c37bf99ec905449